### ISSUE TITLE
Inventory not including all group_vars when using multiple children (or all group)

### ISSUE THREAD

**Issue by stokkie90** [ansible/awx#2574]
##### ISSUE TYPE
 - Bug Report

##### COMPONENT NAME
 - API

##### SUMMARY
Since upgrading to AWX 2.1.0, coming from 2.0.0 the inventory is not working like it should. It should include all group vars where the host is part of, but it only includes the one that it directly configured in.

> group vars are located inside `group_vars/*.yml`

Example hosts file:

```yaml
[parent:children]  <--- this variables are not showing
child_1

[child_1]    <--- this variables are showing 
host_1      <--- this variables are showing
```

##### ENVIRONMENT
* AWX version: 2.1.0
* AWX install method: docker on linux
* Ansible version:  2.7.1
* Operating System: Centos 7
* Web Browser: Chrome

##### STEPS TO REPRODUCE

First configure an inventory that sources from git repo in AWX and use that inventory in AWX and CLI. 

Example:

```bash
# Using Tower inventory 
$ ansible-inventory -i /tmp/tower_inventory.yml --host child_1
{
    "maintenance_reboot": "no", 
    "remote_tower_enabled": "true", 
    "remote_tower_id": 1366
}

# Using Normal inventory 
$ ansible-inventory -i ../inventory/production/ --host child_1 
{
    "env": "prod", 
    "environment": "production", 
    "maintenance_ansible_post_reboot_task": "", 
    "maintenance_ansible_post_update_task": "", 
    "maintenance_ansible_pre_reboot_task": "", 
    "maintenance_ansible_pre_update_task": "", 
    "maintenance_ansible_task": "generic", 
    "maintenance_interval": "weekly", 
    "maintenance_reboot": false, 
    "maintenance_serial": false, 
    "maintenance_strategy": "manual", 
    "maintenance_weekday": "tuesday", 
    "maintenance_window": "night", 
    "sysid_documentation": "none specified", 
    "sysid_environment": "{{ environment }}", 
    "sysid_primary_contact": "Operations"
}

```

##### EXPECTED RESULTS

The above command should give the same output (except the remote_tower variables)

##### ACTUAL RESULTS

See STEPS TO REPRODUCE

##### ADDITIONAL INFORMATION
-

**Comment by AlanCoding** [comment]
Thank you for the fast report on this. The bug is a regression due to the merge of a new feature for job slicing.

I have identified the root cause, which is because of an insufficiently complete definition of "empty" groups when excluding them from the inventory script data. Groups with no hosts were treated as empty, whereas the Ansible core precedent is that only groups with no hosts or groups should be treated as empty. This causes the behavior in your report, because the grandparent group was excluded, which means its variables were not included either.

The fix is very simple, and I hope that it will be able to land today.

**Comment by stokkie90** [comment]
@AlanCoding Thanks for the fast response. I implemented this change. it's now getting the variables from the grandparent, but not the "all" group vars.

**Comment by AlanCoding** [comment]
In your example, were keys from the all group reported in the output of `ansible-inventory -i ../inventory/production/ --host child_1` and which were they?

**Comment by stokkie90** [comment]
Here is the diff:

```
# AWX inventory
$ ansible-inventory -i /tmp/tower_inventory.yml --host child_1
{
    "maintenance_ansible_task": "generic", 
    "maintenance_interval": "weekly", 
    "maintenance_reboot": "no", 
    "maintenance_serial": "yes", 
    "maintenance_strategy": "automatic", 
    "maintenance_weekday": "monday", 
    "maintenance_window": "day", 
    "remote_tower_enabled": "true", 
    "remote_tower_id": 1366
}

# Local inventory
$ ansible-inventory -i ../inventory/production/ --host child_1
{
    "env": "prod", 
    "environment": "production", 
    "maintenance_ansible_post_reboot_task": "", 
    "maintenance_ansible_post_update_task": "", 
    "maintenance_ansible_pre_reboot_task": "", 
    "maintenance_ansible_pre_update_task": "", 
    "maintenance_ansible_task": "generic", 
    "maintenance_interval": "weekly", 
    "maintenance_reboot": false, 
    "maintenance_serial": false, 
    "maintenance_strategy": "manual", 
    "maintenance_weekday": "tuesday", 
    "maintenance_window": "night", 
    "sysid_documentation": "none specified", 
    "sysid_environment": "{{ environment }}", 
    "sysid_primary_contact": "Operations"
}
```

**Comment by stokkie90** [comment]
@AlanCoding I found the problem. the `all_group` is never put back into the `data` dict.

fix should be something like this:
```
        if hostvars:
            data.setdefault('_meta', dict())
            data['_meta'].setdefault('hostvars', dict())
            for host in hosts:
                data['_meta']['hostvars'][host.name] = host.variables_dict
                if towervars:
                    tower_dict = dict(remote_tower_enabled=str(host.enabled).lower(),
                                      remote_tower_id=host.id)
                    data['_meta']['hostvars'][host.name].update(tower_dict)
        data['all'] = all_group
        return data
```

**Comment by AlanCoding** [comment]
I had meant to reproduce your steps fully, but from what you said there, I see clearly what the problem is, and that it lies in the AWX source.

I would suggest that you could also do 

```python
if group_name == 'all':
    continue
```

at

https://github.com/ansible/awx/blob/a338199198f3f246ccd5a6bf45c8e1b6af93f282/awx/main/models/inventory.py#L296

If you want to put in such a PR yourself, I would gladly approve it. A new test, somewhere around here:

https://github.com/ansible/awx/blob/a338199198f3f246ccd5a6bf45c8e1b6af93f282/awx/main/tests/functional/models/test_inventory.py#L41

Would be ideal. If you don't want to write the test I can do it later.

**Comment by AlanCoding** [comment]
For my future reference, the reason why the attempt was made to exclude empty groups in the first place was

https://github.com/ansible/ansible/pull/45621

I'm still not sure if the current criteria for exclusion is right.

**Comment by basos9** [comment]
Hello,
There might be another error condition related to this.

I have parent child groups, and the parent groups seem to not inherit the playbook group variables variables. Note that it works perfect with ansible core.

EDITED: See comment below

**Comment by AlanCoding** [comment]
> I have parent child groups, and the parent groups seem to not inherit the playbook group variables variables.

Before I test the file structure you gave, could you clarify your inventory structure? Parent groups do not inherit variables of their child groups. Also, there is no way for you to directly test group variables in playbooks. So my imagination of the inventory structure is group1->group2->host_a, and you test the variables of host_a inside of your playbook.

**Comment by basos9** [comment]
Ok. I traced it to a feature that I am using which works with ansible and not with AWX. In short I am using extra files to define common parent groups (i.e. groups that are parents of child groups) for different inventories. Moreover these files can be a symlink to one (I tested this is not the issue). 

The benefits of this configuration is that if you have inventories that share common parent groups you define it once. Full example follows.

Structure:
```
inv-prod/
   hosts
   z-common-groups
   group_vars/
     group1.yml
inv-dev/
   hosts
   z-common-groups
   group_vars/
     group1.yml
sys/
   playbook.yml
   group_vars/
     group1.yml <---- this seems to be ignored
```


inv-prod/hosts
```
[subgroup]
host1
```

inv-dev/hosts
```
[subgroup]
host1.dev
```


inv-prod/z-common-groups 
inv-dev/z-common-groups
```
[group1:children]
subgroup
```

inv-prod/group_vars/group1.yml
```
---
  tag: prod
```

inv-dev/group_vars/group1.yml
```
---
  tag: dev
```

sys/group_vars/group1.yml
```
---
  path: "/opt/app"
```

sys/playbook.yml
```
- name: "Deploy something"
  hosts: group1
  tasks:
 
  - command: dosomething tag={{ tag }} path={{ path }} 
```

In this setup awx does not read the group "group1", unless you merge the contents of inv-prod/z-common-groups into inv-prod/hosts. On ansible core this feature works ok, but I am not sure it is documented.

**Comment by AlanCoding** [comment]
when you run this playbook locally via ansible-playbook CLI, what path do you give it for the inventory option, `-i`?

**Comment by basos9** [comment]
I run this playbook like
```
ansible-playbook -i inv-dev sys/playbook.yml
```

**Comment by AlanCoding** [comment]
Thanks, that should be all I need. Testing this use case is on my agenda, but it may be another few days before I get around to it.

**Comment by AlanCoding** [comment]
I tried to reproduce

```
$ tree issues/AWX2574/
issues/AWX2574/
├── inv-dev
│   ├── group_vars
│   │   └── group1.yml
│   ├── hosts
│   └── z-common-groups
├── inv-prod
│   ├── group_vars
│   │   └── group1.yml
│   ├── hosts
│   └── z-common-groups
└── sys
    ├── group_vars
    │   └── group1.yml
    ├── playbook.retry
    └── playbook.yml

6 directories, 9 files
```

![screen shot 2019-03-01 at 4 00 27 pm](https://user-images.githubusercontent.com/1385596/53666140-47fcd600-3c3b-11e9-8e5f-2b7ca280167b.png)

I don't see any problem with how AWX is handling this. That is the same output I got from the CLI.

https://github.com/AlanCoding/Ansible-inventory-file-examples/tree/master/issues/AWX2574

**Comment by basos9** [comment]
Actually it didn't work for me using the repository and playbook from https://github.com/AlanCoding/Ansible-inventory-file-examples/tree/master/issues/AWX2574


![image](https://user-images.githubusercontent.com/1616441/56495566-cf115080-64fe-11e9-965c-e5786f46019b.png)

as the group1 (parent) group is not imported at all.



![image](https://user-images.githubusercontent.com/1616441/56495683-30392400-64ff-11e9-9527-7b5e519ff679.png)

**Comment by AlanCoding** [comment]
All of the needed fixes should be in AWX 4.0.0, what version are you on?
